bin/internal/prepare-environment.sh

```diff
@@ -0,0 +1,181 @@
+# This program and the accompanying materials are made available under the terms of the
+# Eclipse Public License v2.0 which accompanies this distribution, and is available at
+# https://www.eclipse.org/legal/epl-v20.html
+#
+# SPDX-License-Identifier: EPL-2.0
+#
+# Copyright IBM Corporation 2020
+################################################################################
+
+################################################################################
+# This script will prepare all the environment required to start a component.
+#
+# These environment variables should have already been loaded:
+# - INSTANCE_DIR
+#
+# Note: the INSTANCE_DIR can be predefined as global variable, or can be passed
+#       from command line "-c" parameter.
+#
+# Note: this script doesn't rely on <instance-dir>/bin/internal/run-zowe.sh, so
+#       it assumes all environment variables in `instance.env` are not loaded.
+#       With this assumption, we can safely call this script in Zowe Launcher
+#       which doesn't have the environment prepared by run-zowe.sh.
+#
+# Note: this script could be called multiple times during start up.
+################################################################################
+
+if [ "${ZWE_ENVIRONMENT_PREPARED}" = "true" ]; then
+  # environment variables are already prepared, skip running
+  return 0
+fi
+
+# export all variables defined in this script automatically
+set -a
+
+# initialize flag variable to avoid re-run this script
+ZWE_ENVIRONMENT_PREPARED=
+
+# if the user passes INSTANCE_DIR from command line parameter "-c"
+while getopts "c:" opt; do
+  case ${opt} in
+    c) INSTANCE_DIR=${OPTARG};;
+    \?)
+      echo "Invalid option: -${OPTARG}" >&2
+      exit 1
+      ;;
+  esac
+done
+
+# validate INSTANCE_DIR which is required
+if [[ -z ${INSTANCE_DIR} ]]; then
+  echo "INSTANCE_DIR is not defined. You can either pass the value with -c parameter or define it as global environment variable." >&2
+  exit 1
+fi
+
+# find runtime directory
+ROOT_DIR=$(cd $(dirname $0)/../../;pwd)
+
+# prepare some environment variables we always need
+. ${ROOT_DIR}/bin/internal/zowe-set-env.sh
+
+# read the instance environment variables to make sure they exists
+# Question: is there a better way to load these variables since this is already handled by
+#           <instance-dir>/bin/internal/run-zowe.sh
+. ${INSTANCE_DIR}/bin/internal/read-instance.sh
+if [ ! -z "${KEYSTORE_DIRECTORY}" -a -f "${KEYSTORE_DIRECTORY}/zowe-certificates.env" ]; then
+  . ${INSTANCE_DIR}/bin/internal/read-keystore.sh
+fi
+
+# workspace directory variable
+WORKSPACE_DIR=${INSTANCE_DIR}/workspace
+
+# TODO - in for backwards compatibility, remove once naming conventions finalised and sorted #870
+VERIFY_CERTIFICATES="${ZOWE_APIM_VERIFY_CERTIFICATES}"
+
+# zip #1226 - 1.10 -> 1.9 backward compatibility - default keystore type if not supplied
+if [[ -z ${KEYSTORE_TYPE} ]]
+then
+  KEYSTORE_TYPE="PKCS12"
+fi
```

Comment:
Sorry, one extra thing I thought of last night. Since 1.9 (I think it was the 1.12 release) we've posted in the holddata and release notes that zowe-setup-certificates.sh needs to be re-run with an updated .env to pick up the new parameters for SSO, so we should be able to remove this block now

Reply:
I see, it's removed now.